File: tests/test_frontend_modernization_phase1.js
const fs = require("fs");
const path = require("path");
const vm = require("vm");
const assert = require("assert");

const ROOT = path.resolve(__dirname,"..");
const source = fs.readFileSync(path.join(ROOT,"static/js/client-runtime.js"),"utf8");
const template = fs.readFileSync(path.join(ROOT,"templates/index.html"),"utf8");

function buildEnvironment({storageWorks=true}={}){
    const elements = new Map();
    const listeners = {};
    const fetchCalls = [];
    let stateResponse = {ok:true,status:200,requestId:"a".repeat(32)};

    function createElement(){
        return {
            id:"",
            className:"",
            hidden:false,
            textContent:"",
            dataset:{},
            attributes:{},
            setAttribute(name,value){ this.attributes[name] = String(value); }
        };
    }

    const document = {
        readyState:"complete",
        body:{
            appendChild(element){
                if(element.id){ elements.set(element.id,element); }
            }
        },
        createElement,
        getElementById(id){ return elements.get(id) || null; },
        querySelector(selector){
            if(selector === 'meta[name="csrf-token"]'){
                return {content:"csrf-test"};
            }
            return null;
        },
        addEventListener(){}
    };

    function makeStorage(){
        return {
            values:new Map(),
            setItem(key,value){
                if(!storageWorks){ throw new Error("storage blocked"); }
                this.values.set(String(key),String(value));
            },
            removeItem(key){
                if(!storageWorks){ throw new Error("storage blocked"); }
                this.values.delete(String(key));
            }
        };
    }

    async function originalFetch(input,init={}){
        const url = typeof input === "string" ? input : input.url;
        fetchCalls.push({url,init});
        if(url === "/api/client-errors"){
            return {
                ok:true,
                status:202,
                headers:{get(){ return null; }},
                async json(){ return {ok:true,eventId:"server-event"}; }
            };
        }
        if(url === "/api/state" && String(init.method || "GET").toUpperCase() === "PATCH"){
            return {
                ok:stateResponse.ok,
                status:stateResponse.status,
                headers:{get(name){
                    return name === "X-Request-ID" ? stateResponse.requestId : null;
                }},
                async json(){ return {}; }
            };
        }
        return {
            ok:true,
            status:200,
            headers:{get(){ return null; }},
            async json(){ return {}; }
        };
    }

    const window = {
        document,
        location:{origin:"https://tracker.example",pathname:"/app/show/123-private-title"},
        navigator:{onLine:true},
        localStorage:makeStorage(),
        sessionStorage:makeStorage(),
        crypto:{randomUUID(){ return "11111111-2222-3333-4444-555555555555"; }},
        fetch:originalFetch,
        addEventListener(name,handler){
            listeners[name] = listeners[name] || [];
            listeners[name].push(handler);
        },
        setTimeout(){ return 1; },
        clearTimeout(){},
        URL
    };

    const context = {
        window,
        URL,
        console,
        Date,
        Math,
        Promise,
        setTimeout:window.setTimeout,
        clearTimeout:window.clearTimeout
    };
    vm.createContext(context);
    vm.runInContext(source,context,{filename:"client-runtime.js"});

    return {
        window,
        elements,
        listeners,
        fetchCalls,
        setStateResponse(value){ stateResponse = value; }
    };
}

(async()=>{
    assert(template.includes("filename='css/runtime-health.css'"));
    assert(template.includes("filename='js/client-runtime.js'"));
    assert(
        template.indexOf("filename='js/client-runtime.js'") < template.indexOf("filename='js/db.js'"),
        "client runtime observation must install before tracker persistence begins"
    );

    const env = buildEnvironment();
    assert.strictEqual(env.window.TVTrackerClientRuntime.surfaceFromPath("/app/settings/streaming"),"settings");
    assert.strictEqual(env.window.TVTrackerClientRuntime.surfaceFromPath("/app/list/watching"),"tracker");
    assert.strictEqual(env.window.TVTrackerClientRuntime.surfaceFromPath("/app/show/123-private-title"),"detail");

    env.setStateResponse({ok:true,status:200,requestId:"a".repeat(32)});
    await env.window.fetch("/api/state",{method:"PATCH",body:'{"private":"payload"}'});
    const saved = env.elements.get("tv-runtime-save-status");
    assert(saved,"save status should be created");
    assert.strictEqual(saved.textContent,"Saved");
    assert.strictEqual(saved.dataset.state,"saved");

    env.setStateResponse({ok:false,status:503,requestId:"B".repeat(32)});
    await env.window.fetch("/api/state",{method:"PATCH",body:'{"private":"payload"}'});
    assert.strictEqual(saved.textContent,"Save failed — retrying…");
    assert.strictEqual(saved.dataset.state,"error");

    const telemetryCall = [...env.fetchCalls].reverse().find(call=>call.url === "/api/client-errors");
    assert(telemetryCall,"failed save should emit a diagnostic event");
    const diagnostic = JSON.parse(telemetryCall.init.body);
    assert.deepStrictEqual(
        Object.keys(diagnostic).sort(),
        ["category","clientEventId","requestId","status","surface"].sort()
    );
    assert.strictEqual(diagnostic.category,"save");
    assert.strictEqual(diagnostic.surface,"detail");
    assert.strictEqual(diagnostic.status,503);
    assert.strictEqual(diagnostic.requestId,"b".repeat(32));
    const serialized = JSON.stringify(diagnostic);
    for(const forbidden of ["private-title","private","payload","message","stack","url"]){
        assert(!serialized.includes(forbidden),`diagnostic must not include ${forbidden}`);
    }

    const callsBeforeRuntimeError = env.fetchCalls.length;
    assert(env.listeners.error && env.listeners.error.length === 1);
    env.listeners.error[0]({message:"Private title and password=hunter2",error:new Error("secret")});
    await Promise.resolve();
    assert(env.fetchCalls.length > callsBeforeRuntimeError);
    const runtimeCall = env.fetchCalls[env.fetchCalls.length - 1];
    const runtimeDiagnostic = JSON.parse(runtimeCall.init.body);
    assert.strictEqual(runtimeDiagnostic.category,"runtime");
    assert.strictEqual(runtimeDiagnostic.code,"uncaught_error");
    assert(!JSON.stringify(runtimeDiagnostic).includes("hunter2"));
    assert(!JSON.stringify(runtimeDiagnostic).includes("Private title"));

    const blocked = buildEnvironment({storageWorks:false});
    const warning = blocked.elements.get("tv-runtime-warning");
    assert(warning,"blocked persistent storage should produce a visible warning");
    assert(warning.textContent.includes("Browser storage is unavailable"));
    const storageCall = blocked.fetchCalls.find(call=>call.url === "/api/client-errors");
    assert(storageCall,"blocked persistent storage should be reported");
    const storageDiagnostic = JSON.parse(storageCall.init.body);
    assert.strictEqual(storageDiagnostic.category,"storage");
    assert.strictEqual(storageDiagnostic.code,"persistent_storage_unavailable");

    console.log("Frontend modernization phase 1 runtime tests passed.");
})().catch(error=>{
    console.error(error);
    process.exit(1);
});

File: tests/test_frontend_modernization_phase1.py
from __future__ import annotations

import json
import logging
import unittest
from functools import wraps

from flask import Flask, abort, request

from tvtracker.infrastructure.client_errors import (
    CLIENT_ERROR_LIMIT_PER_WINDOW,
    install_client_error_reporting,
)


class _CaptureHandler(logging.Handler):
    def __init__(self) -> None:
        super().__init__()
        self.messages: list[str] = []

    def emit(self, record: logging.LogRecord) -> None:
        self.messages.append(record.getMessage())


class ClientErrorReportingTests(unittest.TestCase):
    def build_app(self, *, clock=lambda: 100.0):
        app = Flask(__name__)
        app.config.update(TESTING=True, SECRET_KEY="test-secret")

        def login_required(view):
            @wraps(view)
            def wrapped(*args, **kwargs):
                if request.headers.get("X-Test-Auth") != "1":
                    return {"ok": False}, 401
                return view(*args, **kwargs)

            return wrapped

        def check_csrf() -> None:
            if request.headers.get("X-CSRF-Token") != "test-csrf":
                abort(403)

        install_client_error_reporting(
            app,
            login_required=login_required,
            check_csrf=check_csrf,
            release_sha="a" * 40,
            clock=clock,
            event_id_factory=lambda: "server-event-id",
        )
        capture = _CaptureHandler()
        app.logger.addHandler(capture)
        app.logger.setLevel(logging.ERROR)
        return app, capture

    def headers(self):
        return {
            "X-Test-Auth": "1",
            "X-CSRF-Token": "test-csrf",
        }

    def test_endpoint_requires_auth_and_csrf(self):
        app, _capture = self.build_app()
        client = app.test_client()
        self.assertEqual(client.post("/api/client-errors", json={}).status_code, 401)
        self.assertEqual(
            client.post(
                "/api/client-errors",
                json={},
                headers={"X-Test-Auth": "1"},
            ).status_code,
            403,
        )

    def test_log_schema_rejects_raw_browser_and_user_data(self):
        app, capture = self.build_app()
        response = app.test_client().post(
            "/api/client-errors",
            json={
                "clientEventId": "client-abc123",
                "category": "save",
                "surface": "tracker",
                "status": 503,
                "requestId": "A" * 32,
                "code": "database_unavailable",
                "message": "Private show title and password=hunter2",
                "stack": "Traceback private data",
                "url": "/app/show/123-private-title",
                "trackerData": {"shows": {"123": {"title": "Private"}}},
                "password": "hunter2",
            },
            headers=self.headers(),
        )
        self.assertEqual(response.status_code, 202)
        self.assertEqual(response.get_json()["eventId"], "client-abc123")
        self.assertEqual(len(capture.messages), 1)
        event = json.loads(capture.messages[0])
        self.assertEqual(
            event,
            {
                "category": "save",
                "clientEventId": "client-abc123",
                "code": "database_unavailable",
                "event": "client_error",
                "releaseSha": "a" * 40,
                "requestId": "a" * 32,
                "status": 503,
                "surface": "tracker",
            },
        )
        serialized = capture.messages[0]
        for forbidden in (
            "Private show title",
            "hunter2",
            "Traceback",
            "/app/show/",
            "trackerData",
            "password",
        ):
            self.assertNotIn(forbidden, serialized)

    def test_invalid_values_collapse_to_safe_defaults(self):
        app, capture = self.build_app()
        response = app.test_client().post(
            "/api/client-errors",
            json={
                "clientEventId": "../../unsafe",
                "category": "password-leak",
                "surface": "/app/show/private-title",
                "status": 999,
                "requestId": "not-a-request-id",
                "code": "contains spaces and secrets",
            },
            headers=self.headers(),
        )
        self.assertEqual(response.status_code, 202)
        event = json.loads(capture.messages[0])
        self.assertEqual(event["clientEventId"], "server-event-id")
        self.assertEqual(event["category"], "runtime")
        self.assertEqual(event["surface"], "app")
        self.assertNotIn("status", event)
        self.assertNotIn("requestId", event)
        self.assertNotIn("code", event)

    def test_oversized_payload_is_rejected_before_logging(self):
        app, capture = self.build_app()
        response = app.test_client().post(
            "/api/client-errors",
            data=json.dumps({"message": "x" * 5000}),
            content_type="application/json",
            headers=self.headers(),
        )
        self.assertEqual(response.status_code, 413)
        self.assertEqual(capture.messages, [])

    def test_log_flood_is_bounded_per_process(self):
        app, capture = self.build_app()
        client = app.test_client()
        for index in range(CLIENT_ERROR_LIMIT_PER_WINDOW):
            response = client.post(
                "/api/client-errors",
                json={"clientEventId": f"event-{index}", "category": "runtime"},
                headers=self.headers(),
            )
            self.assertEqual(response.status_code, 202)
        limited = client.post(
            "/api/client-errors",
            json={"clientEventId": "event-over-limit", "category": "runtime"},
            headers=self.headers(),
        )
        self.assertEqual(limited.status_code, 204)
        self.assertEqual(len(capture.messages), CLIENT_ERROR_LIMIT_PER_WINDOW)


if __name__ == "__main__":
    unittest.main()
